packages/rollup-contracts/contracts/ovm/PartialStateManager.sol

```diff
@@ -0,0 +1,329 @@
+ *         is unlike the FullStateManager which has access to every storage slot.
+ */
+contract PartialStateManager {
+    address constant ZERO_ADDRESS = 0x0000000000000000000000000000000000000000;
+
+    SafetyChecker safetyChecker;
+    FraudVerifier fraudVerifier;
+    ExecutionManager executionManager;
+
+
+    mapping(address=>mapping(bytes32=>bytes32)) ovmContractStorage;
+    mapping(address=>uint) ovmContractNonces;
+    mapping(address=>address) ovmCodeContracts;
+
+    bool public existsInvalidStateAccess;
+    bool public isTransactionSuccessfullyExecuted;
+
+    mapping(address=>mapping(bytes32=>bool)) isVerifiedStorage;
+    mapping(address=>bool) isVerifiedContract;
+    mapping(uint=>bytes32) updatedStorage;
+    uint updatedStorageCounter;
+    mapping(uint=>address) updatedContracts;
+    uint updatedContractsCounter;
+    bytes32 stateRoot;
+
+    modifier preExecution {
+        require(!isTransactionSuccessfullyExecuted, "Function only callable before tx execution.");
+        _;
+    }
+
+    modifier postExecution {
+        require(isTransactionSuccessfullyExecuted, "Function only callable after tx execution.");
+        _;
+    }
+
+    /**
+     * @notice Construct a new FullStateManager with a specified safety checker.
+     */
+    constructor(bytes32 _stateRoot, address _safetyCheckerAddress, address _fraudVerifierAddress) public {
+        stateRoot = _stateRoot;
+        safetyChecker = SafetyChecker(_safetyCheckerAddress);
+        fraudVerifier = FraudVerifier(_fraudVerifierAddress);
+    }
+
+    /**
+     * @notice This is a seperate function because it allows us to first deploy the state manager,
+     * then deploy the execution manager (passing in the state manager address), and then set the execution manager
+     * address in the state manager. This is a bit ugly & probably should be thought through a bit more.
+     */
+    function setExecutionManager(address _executionManagerAddress) public {
+        require(msg.sender == address(fraudVerifier));
+        executionManager = ExecutionManager(_executionManagerAddress);
+    }
+
+    /**
+     * @notice Initialize a new transaction execution
+     */
+    function initNewTransactionExecution() external {
+        require(msg.sender == address(fraudVerifier));
+        existsInvalidStateAccess = false;
+        updatedStorageCounter = 0;
+        updatedContractsCounter = 0;
+    }
+
+    function setIsTransitionSuccessfullyExecuted(bool _isTransactionSuccessfullyExecuted) external {
+        require(msg.sender == address(fraudVerifier));
+        isTransactionSuccessfullyExecuted = _isTransactionSuccessfullyExecuted;
+    }
+
+    function ensureVerifiedStorage(address _ovmContractAddress, bytes32 _slot) private {
+        if (!isVerifiedStorage[_ovmContractAddress][_slot]) {
+            existsInvalidStateAccess = true;
+        }
+    }
+
+    function ensureVerifiedContract(address _ovmContractAddress) private {
```

Comment:
FWIW there is some nuance here where e.g. `ovmEXTCODEHASH` would not require the bytecode itself be verified to be a valid state access.  In fact if we ship that opcode, it's probably necessary to limit fraud proof size, or a contract doing a bunch of these could be problematic.